[Data][LLM] Add should_continue_on_error support for ServeDeploymentStage

nrghosh opened the pull request:
## Description
This is a follow-up to https://github.com/ray-project/ray/pull/59212 which added graceful error handling for vLLMEngineProcessorConfig. 

The limitation to that implementation is that it only applies when the engine shares a lifecycle with the data pipeline - so it doesn't work for when Ray Data is making calls to Serve (DeploymentHandle). So, ServeDeploymentStage also needs error handling. 

## Why is this change needed?

When processing large datasets via Ray Serve deployments, a single malformed
request (e.g., invalid parameters, too-long prompt) would crash the entire
job. Users want the ability to:
1. Continue processing despite individual row failures
2. Receive error information for failed rows for debugging/retry
3. Distinguish between recoverable errors (single row) and fatal errors
   (replica/deployment crash)

## Changes

**ServeDeploymentProcessorConfig**:
- Add `should_continue_on_error: bool = False` field

**ServeDeploymentStageUDF**:
- Add `_generate_with_error_handling()` that catches non-fatal exceptions
  and yields error rows with `__inference_error__` set
- Add `_is_fatal_error()` to distinguish fatal vs recoverable errors
- Include truncated `request_kwargs` in error rows for debuggability

**Fatal Error Detection**:
- `RayActorError`: Replica crashed or unavailable
- `BackPressureError`: System overloaded (note: could benefit from retry)
- `DeploymentUnavailableError`: Deployment failed to deploy
- Also handles `RayTaskError` wrapping by inspecting `.cause`

Non-fatal errors (e.g., ValueError from invalid params) yield error rows
when `should_continue_on_error=True`; fatal errors always propagate.

## Tests

- test_serve_udf_default_raises_on_error
- test_serve_udf_continue_on_error_yields_error_row
- test_serve_udf_mixed_success_and_error
- test_serve_udf_fatal_errors_always_propagate (parametrized for all 3 types)
- test_serve_udf_success_with_continue_on_error_includes_none_error

Related to https://github.com/ray-project/ray/issues/59325
Related to https://github.com/ray-project/ray/pull/52457
Related to https://github.com/ray-project/ray/issues/52449
Fixes https://github.com/ray-project/ray/issues/59411 along the way

nrghosh committed:
Add should_continue_on_error support for ServeDeploymentStage

Signed-off-by: Nikhil Ghosh <nikhil@anyscale.com>

nrghosh committed:
Merge branch 'master' into nrghosh/data-llm-serve-error-handling

nrghosh committed:
Update test to expect should_continue_on_error (Default False)

Signed-off-by: Nikhil Ghosh <nikhil@anyscale.com>

nrghosh committed:
Add integration test for should_continue_on_error with Serve deployment

Signed-off-by: Nikhil Ghosh <nikhil@anyscale.com>

nrghosh committed:
Fix integration test to check __inference_error__ directly

Signed-off-by: Nikhil Ghosh <nikhil@anyscale.com>

nrghosh committed:
Fix: separate error rows before validate_inputs to avoid missing key errors

Fixes #59411. When should_continue_on_error is set, error rows may be missing
required keys (e.g., generated_tokens for DetokenizeUDF). Move error row
separation before validate_inputs so downstream stages skip validation
for error rows.

Save request_kwargs before generate_async pops it

Signed-off-by: Nikhil Ghosh <nikhil@anyscale.com>

nrghosh commented:
/gemini review

richardliaw commented:
Someone else should review but description/behavior makes sense

kouroshHakha commented on python/ray/llm/_internal/batch/stages/serve_deployment_stage.py:
```diff
@@ -111,6 +132,86 @@ async def generate_async(
+                raise
+
+            if not self.should_continue_on_error:
+                raise
```
nit: in one line


`if is_fatal or not should_continue: raise `

kouroshHakha commented on python/ray/llm/_internal/batch/stages/serve_deployment_stage.py:
```diff
@@ -111,6 +132,86 @@ async def generate_async(
+                error_msg,
+            )
+
+            # Include request_kwargs snippet
```
nit: should we create a truncate utility for handling this? we should share it with the other stages?

nrghosh committed:
wip - address comments

- Add shared truncate_str() utility in common.py
- Fix dict spread order bug: **output now comes first so metadata keys
  (request_id, batch_uuid, __inference_error__, etc.) can't be overridden
- Unify udf() code paths: always use _generate_with_error_handling()
  (behavior preserved via should_continue_on_error check inside handler)
- Apply truncate_str to both serve_deployment_stage and vllm_engine_stage

Signed-off-by: Nikhil Ghosh <nikhil@anyscale.com>

nrghosh committed:
wip - invert error handling to allowlist recoverable errors

- Change from allowlisting fatal errors to allowlisting recoverable errors
- Unknown errors now propagate (safer default)
- Add test for unknown error propagation

Signed-off-by: Nikhil Ghosh <nikhil@anyscale.com>

nrghosh committed:
Merge branch 'master' into nrghosh/data-llm-serve-error-handling
